File: src/mk3/PythonPackage/src/tamaf/fiparequestresponder.py
from typing import TYPE_CHECKING
from .fsmbehavior import FSMBehavior
from .simplebehavior import SimpleBehavior
from .performative import Performative

if TYPE_CHECKING:
    from .agent import Agent
    from .aclmessage import ACLMessage
    from .aclmessagetemplate import ACLMessageTemplate

class FIPARequestResponder(FSMBehavior):
    
    def __init__(self, agent: 'Agent', requestTemplate: 'ACLMessageTemplate'):
        super().__init__(agent)
        self.requestTemplate = requestTemplate

        class HandleRequestBehavior(SimpleBehavior):
            def __init__(self, fsm: 'FIPARequestResponder'):
                super().__init__(fsm.agent)
                self.fsm = fsm
                self.transitionCode: int
                self.is_done: bool

            def OnStart(self):
                super().OnStart()
                self.is_done = False
                self.transitionCode = -1

            def Action(self):
                requestMsg = self.Receive(self.fsm.requestTemplate)
                reply = self.fsm.handleRequest(requestMsg)

                if reply is None:
                    return

                if getattr(requestMsg, 'conversationid', None):
                    reply.conversationid = requestMsg.conversationid
                
                self.agent.Send(reply)
                
                if reply.performative == Performative.AGREE:
                    self.fsm.sharedData['request'] = requestMsg
                    self.fsm.sharedData['agreeMessage'] = reply
                    self.transitionCode = 0
                else:
                    self.transitionCode = 1 
                
                self.is_done = True

            def Done(self):
                return self.is_done

            def OnEnd(self):
                return self.transitionCode

        class PrepareResultBehavior(SimpleBehavior):
            def __init__(self, fsm: 'FIPARequestResponder'):
                super().__init__(fsm.agent)
                self.fsm = fsm
                self.is_done: bool

            def OnStart(self):
                super().OnStart()
                self.is_done = False

            def Action(self):
                requestMsg = self.fsm.sharedData['request']
                agreeMsg = self.fsm.sharedData['agreeMessage']
                resultMsg = self.fsm.prepareResultNotification(requestMsg, agreeMsg)

                if resultMsg is None:
                    return

                if getattr(requestMsg, 'conversationid', None):
                    resultMsg.conversationid = requestMsg.conversationid
                self.agent.Send(resultMsg)
                self.is_done = True

            def Done(self):
                return self.is_done

            def OnEnd(self):
                return 0 

        self.AddInitialState(HandleRequestBehavior(self), "handleRequest")
        self.AddState(PrepareResultBehavior(self), "prepareResult")
        
        self.AddTransition("handleRequest", "prepareResult", 0) 
        self.AddTransition("handleRequest", "handleRequest", 1)   
        self.AddTransition("prepareResult", "handleRequest", 0)     

    def handleRequest(self, requestMessage: 'ACLMessage') -> 'ACLMessage':
        pass
        
    def prepareResultNotification(self, requestMessage: 'ACLMessage', responseMessage: 'ACLMessage') -> 'ACLMessage':
        pass

File: src/mk3/JavaPackage/src/main/java/tamaf/FIPARequestResponder.java
package tamaf;

public class FIPARequestResponder extends FSMBehavior {
    protected ACLMessageTemplate requestTemplate;

    public FIPARequestResponder(Agent agent, ACLMessageTemplate requestTemplate) {
        super(agent);
        this.requestTemplate = requestTemplate;

        addInitialState(new HandleRequestBehavior(this), "handleRequest");
        addState(new PrepareResultBehavior(this), "prepareResult");
        
        addTransition("handleRequest", "prepareResult", 0); // Agreed
        addTransition("handleRequest", "handleRequest", 1);   // Refused
        addTransition("prepareResult", "handleRequest", 0);     // Done
    }

    public ACLMessage handleRequest(ACLMessage request) { return null; }
    public ACLMessage prepareResultNotification(ACLMessage request, ACLMessage response) { return null; }

    private class HandleRequestBehavior extends SimpleBehavior {
        FIPARequestResponder fsm;
        int transitionCode;
        boolean isDone;

        @Override
        public void onStart() {
            super.onStart();
            this.transitionCode = -1;
            this.isDone = false;
        }

        HandleRequestBehavior(FIPARequestResponder fsm) { super(fsm.getAgent()); this.fsm = fsm; }
        @Override
        public void action() {
            ACLMessage requestMsg = receive(fsm.requestTemplate);
            if (requestMsg == null) {
                return;
            }
            ACLMessage reply = fsm.handleRequest(requestMsg);

            if (reply == null) {
                return;
            }

            if (requestMsg.getConversationId() != null) {
                reply.setConversationId(requestMsg.getConversationId());
            }
            fsm.getAgent().send(reply);
            if (reply.getPerformative() == Performative.AGREE) {
                fsm.sharedData.put("request", requestMsg);
                fsm.sharedData.put("agreeMessage", reply);
                transitionCode = 0;
            } else {
                transitionCode = 1;
            }
            isDone = true;
        }

        @Override
        public boolean done() { return this.isDone; }

        @Override
        public int onEnd() { return transitionCode; }
    }

    private class PrepareResultBehavior extends SimpleBehavior {
        FIPARequestResponder fsm;
        boolean isDone;

        PrepareResultBehavior(FIPARequestResponder fsm) {
            super(fsm.getAgent());
            this.fsm = fsm;
        }

        @Override
        public void onStart() {
            super.onStart();
            this.isDone = false;
        }

        @Override
        public void action() {
            ACLMessage requestMsg = (ACLMessage) fsm.sharedData.get("request");
            ACLMessage agreeMsg = (ACLMessage) fsm.sharedData.get("agreeMessage");
            ACLMessage resultMsg = fsm.prepareResultNotification(requestMsg, agreeMsg);

            if (resultMsg == null) {
                return;
            }

            if (requestMsg.getConversationId() != null) {
                resultMsg.setConversationId(requestMsg.getConversationId());
            }
            fsm.getAgent().send(resultMsg);
            this.isDone = true;
        }

        @Override
        public boolean done(){
            return this.isDone;
        }

    }
}
